parsers/verilog.c

```diff
@@ -873,12 +873,26 @@ static int _readWordToken (tokenInfo *const token, int c, bool skip)
 	Assert (isWordToken (c));
 
 	clearToken (token);
-	do
+	if (c == '\\')
 	{
-		vStringPut (token->name, c);
-		c = vGetc ();
-	} while (isIdentifierCharacter (c));
-	_updateKind (token);
+		// Escaped identifier (may be empty)
+		c = vGetc (); // skip leading '\'
+		while (isgraph (c))
+		{
+			vStringPut (token->name, c);
+			c = vGetc ();
+		}
```

Comment:
As your comment says, the escaped identifier may be empty, so we need to do special handling at the end :

```c
		if (vStringIsEmpty (token->name) || vStringChar (token->name, 0) == '\\')
		{
			vString *tempName = vStringNewInit ("\\");
			vStringCat (tempName, token->name);
			vStringCopy (token->name, tempName);
			vStringDelete (tempName);
		}
```

Otherwise we will get a prompt `ctags: Notice: ignoring null tag in input.sv(line: 3, language: SystemVerilog)` and miss the `\` tag, consider adding the following test case :

```sv
int \ = 100;
int \\ = 200;
int \\test = 300;
int \\\test = 400;
```

Reply:
That makes sense.  So if the name is empty or starts with `\`, add an additional `\` (even if it's not technically the true name of the variable, but at least it'll be unique).
I'll add that as soon as possible (and the test).

Reply:
Done.  I've also added a test to verify that and `\end` identifier isn't mistaken for an `end` keyword.

Reply:
Nice, your implementation is so concise, LGTM. :thumbsup:

> even if it's not technically the true name of the variable, but at least it'll be unique

In fact, for any non-normal identifier, we need to add an extra `\` when accessing it, unless it can be used as an identifier itself, in which case we can omit `\`. Our current solution is mainly just to implement support for parsing single `\`.

For example, the following variables must be preceded by `\` when used.

```verilog
module test;
  initial begin
    static int \ = 100;
    static int \\ = 200;
    static int \\test = 300;
    static int \\\test = 400;
    static int \r\n = 500;
    static int \test*var = 600;
    static int \**test** = 700;
    static int \begin = 800;
    #1 $display("%d", \ );
    #1 $display("%d", \\ );
    #1 $display("%d", \\test );
    #1 $display("%d", \\\test );
    #1 $display("%d", \r\n );
    #1 $display("%d", \test*var );
    #1 $display("%d", \**test** );
    #1 $display("%d", \begin );
  end
endmodule
```

Only if it is like `\test` or `\var_name`, we can omit `\` in subsequent use.

Reply:
> Nice, your implementation is so concise, LGTM. 👍

Thanks!

> In fact, for any non-normal identifier, we need to add an extra `\` when accessing it, unless it can be used as an identifier itself

I was referring to what the Verilog standard says.  *"Neither the leading backslash character nor the terminating white space is considered to be part of the identifier."*  I.e., `\foo+bar` creates a variable whose "official name" is "foo+bar", without the `\`.  But we probably don't care what Verilog considers to be the "official name" of the variable, and just want a symbol to be able to represent the variable.
(And, as you mentioned, in this case what ctags would give us is what you need to type to access the variable, in case it starts with `\`, so I agree that it's "justified" to use it.)

Reply:
Yes, I totally agree with you. I just made some additional explanations from the compiler's perspective in my previous comment. Anyway, as you said, *"we just want a symbol to be able to represent the variable"*.

In fact, from the perspective of LRM standard alone, there is still an exception. For example, in the latest *IEEE Std 1800-2023*, an example is given for *escaped keyword*.

> `net` is a keyword. `\net` is a user-defined identifier.

That is to say, `\net` creates a variable whose "_official name_" is `\net`, But we don't care.